File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/math/P382_LinkedListRandomNode.java
package com.dsa.leetcode.labuladong.math;

import com.dsa.leetcode.labuladong.linked_list.ListNode;

import java.util.Random;

/**
 * @since 2024/7/27 16:14
 */
public class P382_LinkedListRandomNode {

    class Solution {
        ListNode head;
        Random random;
        public Solution(ListNode head) {
            this.head = head;
            random = new Random();
        }

        public int getRandom() {
            int i = 1, res = 0;
            ListNode cur = head;
            while (cur != null) {
                if (random.nextInt(i) == 0) {
                    res = cur.val;
                }
                i++;
                cur = cur.next;
            }
            return res;
        }
    }



}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/math/P382_LinkedListRandomNode.py
import random
from typing import Optional

from com.dsa.leetcode.labuladong.linked_list import ListNode


class Solution:

    def __init__(self, head: Optional[ListNode]):
        self.head = head

    def getRandom(self) -> int:
        i, res = 1, 0
        cur = self.head
        while cur:
            if random.randrange(i) == 0:
                res = cur.val
            i += 1
            cur = cur.next
        return res
